core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -219,12 +214,12 @@ protected IndexRoutingTable indexRoutingTable(ClusterState clusterState, String
 
     // either routing is set, or type/id are set
 
-    protected IndexShardRoutingTable shards(ClusterState clusterState, String index, String type, String id, String routing) {
-        int shardId = shardId(clusterState, index, type, id, routing);
+    protected IndexShardRoutingTable shards(ClusterState clusterState, String index, String id, String routing) {
+        int shardId = generateShardId(clusterState, index, id, routing);
         return shards(clusterState, index, shardId);
     }
 
-    protected IndexShardRoutingTable shards(ClusterState clusterState, String index, int shardId) {
+    public IndexShardRoutingTable shards(ClusterState clusterState, String index, int shardId) {
```

Comment:
I added a shardRoutingTable shortcut to RoutingTable, which returns the shard routing table and throws appropriate exceptions when the index and/or shard id is unknown in https://github.com/areek/elasticsearch/commit/688197f0be6050c4a63bd4cb71de88cc7dd29524.